python_modules/dagster/dagster/_core/external_execution/task.py

```diff
@@ -57,228 +56,224 @@ def __init__(
+        path = self._prepare_io_path(self._output_path, "output")
+        env = {DAGSTER_EXTERNAL_ENV_KEYS["output"]: path}
+        if not os.path.exists(path):
+            os.mkfifo(path)
+        # We open a write file descriptor for a FIFO in the orchestration
+        # process in order to keep the FIFO open for reading. Otherwise, the
+        # FIFO will receive EOF after the first file descriptor writing it is
+        # closed in an external process. Note that `O_RDWR` is required for
+        # this call to succeed when no readers are yet available, and
+        # `O_NONBLOCK` is required to prevent the `open` call from blocking.
+        dummy_handle = os.open(path, os.O_RDWR | os.O_NONBLOCK)
+        thread = self._default_output_thread(path)
+        yield None, env
+        os.close(dummy_handle)
+        thread.join()
+        os.remove(path)
+
+    # Socket server is started/shutdown in a dedicated context manager to share logic between input
+    # and output.
+    @contextmanager
+    def socket_output(self) -> Iterator[OutputIOYield]:
+        env = {
+            DAGSTER_EXTERNAL_ENV_KEYS["host"]: (
+                self._socket_server_host or DAGSTER_EXTERNAL_DEFAULT_HOST
+            ),
+            DAGSTER_EXTERNAL_ENV_KEYS["port"]: str(
+                self._socket_server_port or DAGSTER_EXTERNAL_DEFAULT_PORT
+            ),
+        }
+        yield None, env
+
+    @contextmanager
+    def _socket_server_context_manager(self) -> Iterator[None]:
+        host = self._socket_server_host or DAGSTER_EXTERNAL_DEFAULT_HOST
+        port = self._socket_server_port or DAGSTER_EXTERNAL_DEFAULT_PORT
+        sockaddr = (host, port)
+        thread = self._start_socket_server_thread(sockaddr)
+        yield
+        self._shutdown_socket_server(sockaddr)
+        thread.join()
 
     # ########################
     # ##### THREADED IO ROUTINES
     # ########################
 
-    def _default_input_thread(self, write_target) -> Thread:
-        return Thread(target=self._write_input, args=(write_target,), daemon=True)
+    def _start_socket_server_thread(self, sockaddr: SocketAddress) -> Thread:
+        is_server_started = Event()
+        thread = Thread(
+            target=self._start_socket_server, args=(sockaddr, is_server_started), daemon=True
+        )
+        thread.start()
+        is_server_started.wait()
+        return thread
 
-    def _default_output_thread(self, read_target) -> Thread:
-        return Thread(target=self._read_output, args=(read_target,), daemon=True)
+    def _default_input_thread(self, path_or_fd: Union[str, int]) -> Thread:
+        thread = Thread(target=self._write_input, args=(path_or_fd,), daemon=True)
+        thread.start()
+        return thread
+
+    def _default_output_thread(self, path_or_fd: Union[str, int]) -> Thread:
+        thread = Thread(target=self._read_output, args=(path_or_fd,), daemon=True)
+        thread.start()
+        return thread
```

Comment:
I must have missed this. Why do we need threads at all in this system?

Reply:
`Popen` is designed to be non-blocking.  I also think writing a non-blocking variant will be much closer to what we want for REST API-driven calls.